### DISCUSSION TITLE
pg_temp.tipg_catalog takes ~1 minute when there are large spatial tables

### DISCUSSION THREAD

**Discussion by hrodmn** [developmentseed/tipg#139] [Q&A] [answered]
I have been using TiPG deployed as a Lambda function via the `eoapi-cdk` constructs and it is awesome! However, as the size of my spatial tables increases, the FastAPI application startup time exceeds the default timeout for the Lambda function (30 seconds). I am able to use TiPG successfully on the same database locally via `uvicorn`, but I can see that the application takes a minute to reach an active state.

My largest table has 3.6 million polygon geometries in it and has foreign keys on two other tables. On an identical database with only a few thousand records in that table the startup time is about 1 second! Has my deployment outgrown the Lambda option or if there is something I need to configure differently in TiPG (or even my database schema) to improve application performance? If I was only running the startup occasionally (i.e. on an always-on server) I don't think I would have a problem, but I would like to keep using Lambda if possible since we are not using the service very heavily right now.

I did some local debugging and can see that the `pg_temp.tipg_catalog` function is taking almost all of the startup time.

**Comment by vincentsarago** [comment, marked as answer]
@hrodmn In theory we could remove the internal catalog to list the collections `dynamically` (previous discussion: https://github.com/developmentseed/tipg/issues/37) 

As of today, we don't have $ nor time to work on this.

The main issue might come from https://github.com/developmentseed/tipg/blob/e921266dfb1b19ebc012d7533e860faa2b3f5623/tipg/sql/dbcatalog.sql#L51-L88

Where we try to get the spatial and temporal extents for each table! This should be fairly quick if there are indexes on both but maybe not. 

Maybe we could add a settings to turn ON/OFF this step 

cc @bitner

**Reply by vincentsarago** [reply]
note on `stateless catalog`, https://github.com/developmentseed/tipgstac is a demo application showing how we can work directly without the catalog

**Reply by hrodmn** [reply]
@vincentsarago thank you as always for the suggestions! Right now I do not have indexes set on either the geometry OR temporal columns in my spatial tables, so I will test that out first.

**Reply by vincentsarago** [reply]
I've an issue opened in eoAPI https://github.com/developmentseed/eoAPI/issues/68

**Reply by hrodmn** [reply]
@vincentsarago your theory about the bottleneck is correct.

I added indexes to the spatial and temporal columns in my tables and was able to reduce the runtime for `pg_catalog` from ~70 seconds to ~40 seconds, so that does make a difference! However, that probably isn't good enough for running `tipg` as a Lambda function.

I commented out the lines that calculate the extents of the spatial and temporal columns and that brought startup time down to one second, so, for the purposes of running `tipg` in a serverless application, maybe a toggle to skip that step would be really useful.

Would I lose any functionality by skipping the extent calculations?

**Reply by vincentsarago** [reply]
I think it will make sense

We had a `datatime_extent` boolean in the past https://github.com/developmentseed/tipg/blob/main/tipg/settings.py#L53 which is not used any more but I think we could add both 

`datetime_extent` and `spatial_extent` options in https://github.com/developmentseed/tipg/blob/main/tipg/settings.py#L151-L164

cc @bitner

**Reply by hrodmn** [reply]
I will take a crack at the changes this afternoon - PR incoming!

**Reply by jr-elliott** [reply]
@hrodmn did you notice any functionality loss when you disabled the extents?

**Reply by hrodmn** [reply]
@jr-elliott if you disable the spatial/temporal extent calculation each `collection` object will lack a spatial and/or temporal extent. This doesn't break anything because the spatial and temporal extents are optional according to the [OGC Features API spec](https://docs.ogc.org/is/17-069r4/17-069r4.html#_requirements_class_core):

> - An optional extent that can be used to provide an indication of the spatial and temporal extent of the collection - typically derived from the data;

If you are running tipg on a database with really large tables in a serverless runtime environment I highly recommend setting `TIPG_DB_TEMPORAL_EXTENT=FALSE` and `TIPG_DB_SPATIAL_EXTENT=FALSE` to avoid an expensive application startup process.
